### ISSUE TITLE
Travis CI errors (for non-master branch)

### ISSUE THREAD

**Issue by vojtechhuser** [OHDSI/Achilles#154]
To test the code (e.g., cast of date into varchar) across 4 DB engines, Travis CI tries several servers.

This testing is currently failing.

**Comment by vojtechhuser** [comment]
I think this is due to connection details not correct. The log says:


> > library(testthat)
> > library(Achilles)
> Loading required package: SqlRender
> Loading required package: DatabaseConnector
> Loading required package: RJDBC
> Loading required package: DBI
> Loading required package: rJava
> Loading required package: rjson
> > 
> > test_check("Achilles")
> Connecting using PostgreSQL driver
> Error in connect(dbms = connectionDetails$dbms, user = connectionDetails$user,  : 
>   Error: database name not included in server string but is required for PostgreSQL. Please specify server as <host>/<database>
> 1. Error: Achilles main does not throw an error on Postgres (@test-main-achilles.R#23) 
> object 'result' not found
> 1: expect_true(class(result) == "achillesResults") at testthat/test-main-achilles.R:23
> 2: expect(identical(as.vector(object), TRUE), sprintf("%s isn't true.", lab), info = info)
> 3: as.expectation(exp, ..., srcref = srcref)
> 4: identical(as.vector(object), TRUE)
> 5: as.vector(object)
> 
> Connecting using SQL Server driver
> Error in .jcall(drv@jdrv, "Ljava/sql/Connection;", "connect", as.character(url)[1],  : 
>   com.microsoft.sqlserver.jdbc.SQLServerException: The TCP/IP connection to the host localhost, port 1433 has failed. Error: "Connection refused. Verify the connection properties. Make sure that an instance of SQL Server is running on the host and accepting TCP/IP connections at the port. Make sure that TCP connections to the port are not blocked by a firewall.".
> 2. Error: Achilles main does not throw an error on SQL Server (@test-main-achilles.R#42) 
> object 'result' not found
> 1: expect_true(class(result) == "achillesResults") at testthat/test-main-achilles.R:42
> 2: expect(identical(as.vector(object), TRUE), sprintf("%s isn't true.", lab), info = info)
> 3: as.expectation(exp, ..., srcref = srcref)
> 4: identical(as.vector(object), TRUE)
> 5: as.vector(object)
> 
> Connecting using Oracle driver
> - using THIN to connect
> Error in rJava::.jcall(drv@jdrv, "Ljava/sql/Connection;", "connect", as.character(url)[1],  : 
>   java.sql.SQLRecoverableException: IO Error: The Network Adapter could not establish the connection
> - Trying using TNSName
> Error in rJava::.jcall(drv@jdrv, "Ljava/sql/Connection;", "connect", as.character(url)[1],  : 
>   java.sql.SQLRecoverableException: IO Error: Connect identifier was empty.
> 3. Error: Achilles main does not throw an error on Oracle (@test-main-achilles.R#62) 
> object 'result' not found
> 1: expect_true(class(result) == "achillesResults") at testthat/test-main-achilles.R:62
> 2: expect(identical(as.vector(object), TRUE), sprintf("%s isn't true.", lab), info = info)
> 3: as.expectation(exp, ..., srcref = srcref)
> 4: identical(as.vector(object), TRUE)
> 5: as.vector(object)
> 
> Connecting using PostgreSQL driver
> Error in connect(dbms = connectionDetails$dbms, user = connectionDetails$user,  : 
>   Error: database name not included in server string but is required for PostgreSQL. Please specify server as <host>/<database>
> 4. Failure: Achilles export does not throw an error on Postgres (@test-main-achilles.R#80) 
> file.exists("postgresql/dashboard.json") isn't true.
> 
> 
> Connecting using Oracle driver
> - using THIN to connect
> Error in rJava::.jcall(drv@jdrv, "Ljava/sql/Connection;", "connect", as.character(url)[1],  : 
>   java.sql.SQLRecoverableException: IO Error: The Network Adapter could not establish the connection
> - Trying using TNSName
> Error in rJava::.jcall(drv@jdrv, "Ljava/sql/Connection;", "connect", as.character(url)[1],  : 
>   java.sql.SQLRecoverableException: IO Error: Connect identifier was empty.
> 5. Failure: Achilles export does not throw an error on Oracle (@test-main-achilles.R#116) 
> file.exists("oracle/dashboard.json") isn't true.
> 
> 
> testthat results ================================================================
> OK: 0 SKIPPED: 0 FAILED: 5
> 1. Error: Achilles main does not throw an error on Postgres (@test-main-achilles.R#23) 
> 2. Error: Achilles main does not throw an error on SQL Server (@test-main-achilles.R#42) 
> 3. Error: Achilles main does not throw an error on Oracle (@test-main-achilles.R#62) 
> 4. Failure: Achilles export does not throw an error on Postgres (@test-main-achilles.R#80) 
> 5. Failure: Achilles export does not throw an error on Oracle (@test-main-achilles.R#116) 
> 
> Error: testthat unit tests failed
> Execution halted
> 
> 
> Done. Your build exited with 1.

**Comment by t-abdul-basser** [comment]
Yes. I have asked @chris and @lee about this. It may be possible for modify the export variables to control this behavior.

**Comment by chrisknoll** [comment]
Sorry @t-abdul-basser , @vojtechhuser , I have no knowledge on how the test environment was set up, what the credentials are to the CI environment's SQL/Postgresql/Oracle test servers, so I can not offer any help with what might be wrong with the CI environment and settings.

This is something that greatly deserves a wiki page, and if it already exists, I apologize, I don't think I've seen it or know of its existence.

**Comment by t-abdul-basser** [comment]
Thanks @chrisknoll. I am looking into this and will ask @msuchard if I have questions :)

**Comment by schuemie** [comment]
As I mentioned in another issue, I'm pretty sure this problem is caused **because the environmental variables are not available in tests of pull requests**. 

I can see two solutions:
- Merge the pull request with a branch other than master, and have Travis test it there
- Someone who has the environmental variables set locally (e.g. Chris) can fetch the pull request and test it locally

**Comment by t-abdul-basser** [comment]
Thanks @schuemie (!= @suchard: sorry early morning) . Yes I took note of your comment and planned to explore with it in mind. Quick comment re your suggestions though
* From a previous quick glance of the Travis build history it seemed that similar failures occurred on *commits to branches not just PR* but perhaps I am mistaken
* Running locally and manually  != CI so I would prefer to avoid falling back to that.
I will explore and follow up.

**Comment by chrisknoll** [comment]
I agree with @t-abdul-basser about running this locally.  I feel like it's just a matter of the travis CI process to assert those env variables before executing the test (just like a dev would need to do after checking out the branch locally).  I'm not understanding the technical limitation where the CI execution can't assert those env variables that are specific to the CI environment.  It shouldn't matter about which branch is checked out (which is essentially what a PR is) it should all be executed in the same CI container.  Please correct me if I'm misunderstanding anything here.

**Comment by vojtechhuser** [comment]
Thank you for inputs to everyone.
Martijn's suggestion is interesting 

> Merge the pull request with a branch other than master, and have Travis test it there

Let me simplify the issue even further. 
TravisCI fails for the master branch. (as of now).

https://travis-ci.org/OHDSI/Achilles/branches

see here
![image](https://cloud.githubusercontent.com/assets/7526119/20144543/ea182822-a66a-11e6-85e6-5fd7d5d28eae.png)

**Comment by schuemie** [comment]
Yes, master is failing for an entirely different reason, as discussed in [this issue](https://github.com/OHDSI/Achilles/pull/152).

**Comment by vojtechhuser** [comment]
I renamed the issues to separate the two travisCI problems. (env. variables, and long join)

This thread only relates to non master branch errors.

Chris wrote:
> I feel like it's just a matter of the travis CI process to assert those env variables before executing the test 
(just like a dev would need to do after checking out the branch locally).

Just additional detail:
Several months ago - travisCI was working fine and it was able to test several DBMSs (using the test R code). I had some devH1 and devH2 branches and those were working just fine (I was able to demostrate that the new features are OK).  I am hoping that @t-abdul-basser  will be able to make it work again (e.g., for devH8 branch).  (or even for a forked branche githubuserXYZ/Achilles [as wiki page dictates to use]).

**Comment by vojtechhuser** [comment]
Thank you everyone for looking into this. 

New input from Chris is:


it looks like we're not seeing the exports happening in that build log:
https://travis-ci.org/OHDSI/Achilles/builds/173999526#L1397

it's missing this content at the top of the log (this is from this PR's build log):

Setting environment variables from repository settings

> $ export CDM5_POSTGRESQL_USER=[secure]
> $ export CDM5_POSTGRESQL_PASSWORD=[secure]
> $ export CDM5_POSTGRESQL_SERVER=[secure]
> $ export CDM5_POSTGRESQL_CDM_SCHEMA=[secure]

The hint here is that it's pulling it from the 'repository settings'. Maybe we could have the TRAVIS-CI build define those exports explicitly in the task instead of reading from a repository setting because maybe forked repos don't have that? In any case, without those exports, the tests will fail. So could we also address something in the Travis jobs where when a test is going to execute, that the set of necessary travis-ci env vars are exported?

**Comment by msuchard** [comment]
Unfortunately, encrypted environmental variables are *not* available to pull requests from different repositories (to keep evil users from doing `printenv`).    For pull-requests, we need a new check mechanisms that does not require encryption.

**Comment by vojtechhuser** [comment]
We had the problem of devBranch vs myRepo/Achilles about 10 months ago. 
We had to go with devBranches.  (early 2016)

I was saying that to @t-abdul-basser  in october 2016 - but I guess a ownRepo approach had stronger advocates. 

Taha and Chris, just issue at some point specific development guidance just for Achilles. 

(even it is contradicts the ohdsi wiki stated standards).

**Comment by t-abdul-basser** [comment]
Good point @vojtechhuser. I realize that there is a history here and that I am a Achilles newbie :) so I will start as per your suggestion by mentioning what I mentioned elsewhere earlier: I would suggest that developers develop on forks, submit PRs to OHDSI/Achilles feature branches, that these feature/rc branches be merge into master, etc. Open to suggestions though @chrisknoll

**Comment by chrisknoll** [comment]
External repo PRs should be supported (and is to be expected as the norm because not everyone is going to be an internal collaborator to the repo, rather the'll fork->pr).

I understand what you are saying @msuchard but what I can't get my mind around is that when the travis-ci process kicks up, i'd expect all the env vars defined here:
https://travis-ci.org/OHDSI/Achilles/settings

would still be avilable in the context of the Travis CI job.  So i'm sorry I'm being dense on this point, but it doesn't make sense to me, woudln't anyone be able to put in a printenv in a local branch anyways? 

In any case, the next order of business is how to make the credentials of the 3 test DBs available during the build process regardless of where the build is comming from (a push, pr locally, pr from remote repo) and i'll open up a new issue on that.

**Comment by chrisknoll** [comment]
This can be closed and the approach to deal with merging external PRs is in #165 discussion.
